### ISSUE TITLE
Implementing New Operators

### ISSUE THREAD

**Issue by II-Matto** [apache/incubator-mxnet#9012]
I would like to implement a new operator with MXNet, and the operator is already implemented with Caffe. What is the easiest way to port my Caffe implementation to MXNet?

The comment https://github.com/apache/incubator-mxnet/issues/7652#issuecomment-325894475 says that it is easier to implement complex operators using the legacy interface. But will the **legacy interface** be unsupported in the future? If it is OK to use the legacy interface, can someone provide links to some **comprehensive tutorials or examples**?

-------

To use the new nnvm interface for defining operators, I read this document: [A Beginner's Guide to Implementing Operators in MXNet Backend](https://github.com/apache/incubator-mxnet/blob/master/docs/how_to/add_op_in_backend.md). The step-by-step explanations are clear and I have understood the general pipeline of defining a new operator, but there still seems to be some obstacles in the way.

1. How can I define **_variables storing intermediate results_** which can be used in both forward and backward computations? These variables will act like class member variables (assuming the operator is defined as a class). This is similar to the question in the comment https://github.com/apache/incubator-mxnet/issues/7652#issuecomment-338910634.
2. How can I define a _**operator with learnable parameters**_ (e.g. convolution operator)? The current implementations of operators such as convolution seem to use a different (and legacy) interface. And the [tensor examples](https://github.com/apache/incubator-mxnet/tree/master/src/operator/tensor) are all parameter-free operators like dot product.
3. Where can I find _**explanations of the various macros**_ to be used in operator definitions? The many macros are very confusing and it is hard for me to determine when I should use what and how. It would be very helpful if some templates for typical usages can be provided and explained.
4. How can I **_correctly reference the headers_** in [src/operator](https://github.com/apache/incubator-mxnet/tree/master/src/operator), if I put my operator definition files in an independent directory outside MXNet? Many source files simply use relative paths, e.g. `#include #include "../mxnet_op.h"` and `#include "../mshadow_op.h"`, which are not suitable if a file is located outside MXNet directories.
5. How can I **_specify multiple paths for the_** `EXTRA_OPERATORS` in  [config.mk](https://github.com/apache/incubator-mxnet/blob/master/make/config.mk)? After checking the [Makefile](https://github.com/apache/incubator-mxnet/blob/master/Makefile) and [CMakeLists.txt](https://github.com/apache/incubator-mxnet/blob/master/CMakeLists.txt), I find that only one path is supported (Please correct me if I am wrong).

**Comment by piiswrong** [comment]
@reminisce

**Comment by reminisce** [comment]
1. You can take a look at this. That PR refactored most of nn operators into using nnvm interfaces. You can register that attribute for an operator to share states. https://github.com/apache/incubator-mxnet/pull/8302/files#diff-c7e5df696274a535051595cc84f35353R444
2. You can take a look at the convolution refactored code which shows how to define learnable parameters. https://github.com/apache/incubator-mxnet/pull/8302/files#diff-be97f478f0c55a9c72e3bcb2119bf006R413
3. What macros are you referring to? I don't what IDE or editor you are using for writing C++ code, but I think most of the main stream ones should support macro definition navigation. Let me know the specific macros you could not find the definitions for.
4. If you want to keep your operator implementation separated from the existing MXNet code base, you can create a folder for you only under the `src/operator`.
5. @cjolivier01

**Comment by II-Matto** [comment]
@piiswrong @reminisce Thanks for your replies!

1. As for the shared states,
    1. Can multiple `FMutateInputs` be used? Or should I just return multiple variables for one `FMutateInputs`? Can you provide an example?
    2. How can the registered `FMutateInputs` be used? Currently I haven't found the position in batch normalization implementation files where the [`std::vector<uint32_t>{3, 4}`](https://github.com/apache/incubator-mxnet/pull/8302/files?diff=split#diff-c7e5df696274a535051595cc84f35353R445) is used.
    3. What if the size of the variable storing intermediate results can only be determined during the computation on the fly? In such cases, I will need to compute the size of the variable and allocate the space dynamically (`Blob::Reshape()` in Caffe). And I will need to switch between CPU and GPU storage during the computation (`Blob::cpu_data()` and `Blob::gpu_data()` in Caffe), since the values stored at GPU can not be directly used in host code.
    4. BTW, why the [`inputs()`](https://github.com/apache/incubator-mxnet/pull/8302/files#diff-6e6adac74a85b946a9a8ac6e5875e20aR232) in the `BatchNormGradCompute()` have the length of 11 and the [`in_data_start`](https://github.com/apache/incubator-mxnet/pull/8302/files#diff-6e6adac74a85b946a9a8ac6e5875e20aR235) equals 3?
2. As for operator with learnable parameters,
    1. I have checked out the convolution implementation files. The only thing I find is that the `weight` and `bias` are treated similarly to the `data` as one of the operator inputs. But what exactly is it that distinguishes between input data and operator parameters (so that the program knows the parameters need to be stored and updated)?
    2. In the convolution implementation, `FResourceRequest` is used to claim one extra resource needed. If I need multiple resources, what should I do? Should I just return a longer resource request vector and define more values in the `ConvolutionOpResource` enum as below?
```
enum ConvolutionOpResource { kTempSpace1, kTempSpace2, kTempSpace2 };
...
.set_attr<FResourceRequest>("FResourceRequest", [](const NodeAttrs& n) {
  return std::vector<ResourceRequest>{ ResourceRequest::kTempSpace, ResourceRequest::kTempSpace, ResourceRequest::kTempSpace };
})
```
3. As for the macros, I can checkout their definitions on my own. But I hope there are some clear explanations and examples telling me when I should use them. This allows users to start to develop their own code more quickly. In Caffe, I know I only need to use one `CUDA_KERNEL_LOOP` macro in most cases. But in MXNet, I feel totally blank, not knowing whether I should look for and use macros. It would be low efficient if each developer needs to examine all the macro definitions on their own.
    - The [beginner guide](https://github.com/apache/incubator-mxnet/blob/master/docs/how_to/add_op_in_backend.md) shows usages of `MSHADOW_TYPE_SWITCH`, `MXNET_ASSIGN_REQ_SWITCH` and `KERNEL_ASSIGN`. The former two can be easily understood, while the last one seems to only suit simple computations. In more complex computations, one will have pre-processings (e.g. computing sizes of intermediate variables and allocate memory dynamically), multiple callings of different CUDA kernel functions, post-processings, etc.
4. Can my own code be put in completely separate locations? For example, the MXNet codebase lies at `<root>/<sub_root_1>/mxnet`, and I want to put my own code at `<root>/<sub_root_2>/<my_own_code_dir>`.
5. Hi, @cjolivier01, could you help me with this?

**Comment by reminisce** [comment]
1.
i-ii You can take a look at this line: https://github.com/apache/incubator-mxnet/pull/8302/files#diff-8417803f764246e7864dc1d54c256dcfR661. It tells the indices of aux data in the input data vector as registered in the form of `vector<uint32_t>{3, 4}`.
iii MXNet only supports dynamically allocating memory space for sparse tensors in the format of row-sparse and csr.
iv I guess inputs[1] and inputs[2] stores the grads of `mean` and `var`. Does it explain why there are 11 inputs to you?
2. 
i weights as a param in convolution just specifies the shape of filter. The graph executor will allocate a tensor for it. Once you define the backward function for a input tensor, it becomes learnable.
ii Right now, MXNet only supports allocating temp space once per forward/backward function. You can pre-calculate the total space you need and allocate once for all.
3. Agree that too many macros lower the readability of the code and makes running gdb tracing harder. But the ones you listed are mostly the common ones you are going to see while writing cpp code and they let you escape from writing quite a lot of boilerplate code.
4. I guess so. You just need to write a correct Makefile to include headers and link the MXNet lib.

**Comment by TaoLv** [comment]
For 1.iv:
The inputs of BatchNormGradCompute can be listed as below:
[grad wrt bn output data,
 grad wrt bn output mean,
 grad wrt bn output variance,
 input data to bn forward function,
 input gamma to bn forward function,
 input beta to bn forward function,
 moving mean to bn forward function,
 moving variance to bn forward function,
 output data of bn forward function,
 output mean of bn forward function,
 output variance of bn forward function].
Some of them may be NULL according to different parameters. But I am not sure if the input order is same for all operators. @reminisce, Can you comment on that? I don't konw where does mxnet decide how many inputs are needed by a backward function or what is the input order looks like?

**Comment by TaoLv** [comment]
Seems defined in this sentence:
`.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseInOut{"_backward_BatchNorm"})`

ElemwiseGradUseInOut decides the input number and order.
